core/src/main/java/com/netflix/conductor/core/execution/tasks/Wait.java

```diff
@@ -12,29 +12,120 @@
+
+import org.apache.commons.lang3.time.DateUtils;
+import org.apache.logging.log4j.util.Strings;
 import org.springframework.stereotype.Component;
 
 import com.netflix.conductor.core.execution.WorkflowExecutor;
 import com.netflix.conductor.model.TaskModel;
 import com.netflix.conductor.model.WorkflowModel;
 
 import static com.netflix.conductor.common.metadata.tasks.TaskType.TASK_TYPE_WAIT;
-import static com.netflix.conductor.model.TaskModel.Status.IN_PROGRESS;
+import static com.netflix.conductor.model.TaskModel.Status.*;
 
 @Component(TASK_TYPE_WAIT)
 public class Wait extends WorkflowSystemTask {
 
+    public static final String DURATION_INPUT = "duration";
+    public static final String UNTIL_INPUT = "until";
+    public static final String TIMEOUT = "timeout";
+
+    private static final String[] patterns =
+            new String[] {"yyyy-MM-dd HH:mm", "yyyy-MM-dd HH:mm z", "yyyy-MM-dd"};
+
     public Wait() {
         super(TASK_TYPE_WAIT);
     }
 
+    private static Duration parseDuration(String text) {
+        Matcher m =
+                Pattern.compile(
+                                "\\s*(?:(\\d+)\\s*(?:days?|d))?"
+                                        + "\\s*(?:(\\d+)\\s*(?:hours?|hrs?|h))?"
+                                        + "\\s*(?:(\\d+)\\s*(?:minutes?|mins?|m))?"
+                                        + "\\s*(?:(\\d+)\\s*(?:seconds?|secs?|s))?"
+                                        + "\\s*",
+                                Pattern.CASE_INSENSITIVE)
+                        .matcher(text);
+        if (!m.matches()) throw new IllegalArgumentException("Not valid duration: " + text);
+
+        int days = (m.start(1) == -1 ? 0 : Integer.parseInt(m.group(1)));
+        int hours = (m.start(2) == -1 ? 0 : Integer.parseInt(m.group(2)));
+        int mins = (m.start(3) == -1 ? 0 : Integer.parseInt(m.group(3)));
+        int secs = (m.start(4) == -1 ? 0 : Integer.parseInt(m.group(4)));
+        return Duration.ofSeconds((days * 86400) + (hours * 60L + mins) * 60L + secs);
+    }
+
+    private static Date parseDate(String date) throws ParseException {
+        return DateUtils.parseDate(date, patterns);
+    }
+
     @Override
     public void start(WorkflowModel workflow, TaskModel task, WorkflowExecutor workflowExecutor) {
+
+        String duration =
+                Optional.ofNullable(task.getInputData().get(DURATION_INPUT)).orElse("").toString();
+        String until =
+                Optional.ofNullable(task.getInputData().get(UNTIL_INPUT)).orElse("").toString();
+
+        if (Strings.isNotBlank(duration) && Strings.isNotBlank(until)) {
+            task.setReasonForIncompletion(
+                    "Both 'duration' and 'until' specified. Please provide only one input");
+            task.setStatus(FAILED);
+            return;
+        }
+
+        if (Strings.isNotBlank(duration)) {
+
+            Duration timeDuration = parseDuration(duration);
+            task.getOutputData()
```

Comment:
Instead of adding the `timeout` to the output data, we suggest that its added as a field to the `TaskModel`. We think its a better fit because,

1. `TaskModel` is internal only and fields in it are not exposed to the user. Adding it to outputData makes the `timeout` field part of the task contract.
2. We have plans to create custom TaskModels which only contain fields for a particular task type.

Reply:
updated to use TaskModel